Handle quotes in more commands

matttbe opened the pull request:
A follow-up of #287 from @mhartmay: different fixes and one small feature.

This PR would close #265 and #286

mhartmay commented on virtme_ng/run.py:
```diff
@@ -874,23 +872,20 @@ def make(self, args):
             cross_compile = None
             cross_arch = None
-        make_command = MAKE_COMMAND
+        make_command = MAKE_COMMAND.split()
```
Please use `shlex.split` or even better, change MAKE_COMMAND to an array, but please do not forget to copy the value before changing.

```
MAKE_COMMAND = [...]

...
make_command = MAKE_COMMAND.copy()
```

mhartmay replied:
Yep.

matttbe replied:
thanks, that's what I did: https://github.com/arighi/virtme-ng/compare/14a20c7d06120d24b6d8692e0ece19daeaabc5f8..dc05350469d80bd056dbbcc2134f49b37d4fd9bb

mhartmay reviewed (commented):
Hmmm `args.envs` is no good name then, is it?

And I'm not sure if this can be used in a different way than you intended. e.g. in the README.md there is an example how you can pass makefile variables to the makefile call `vng O=.virtme/build --verbose`.

So what will happen if you use (I did not test):

`vng echo --ssh-client test`

Update: Ideally, we can get rid of `_format_cmd`.

matttbe commented:
> Hmmm `args.envs` is no good name then, is it?

Indeed, but it looks like it is the same name in other places, and it was already used as a non "env var" in different places, like to execute a command:

    vng -- uname -r

Anyway, we can change, not sure what would be the better name.

> And I'm not sure if this can be used in a different way than you intended. e.g. in the README.md there is an example how you can pass makefile variables to the makefile call `vng O=.virtme/build --verbose`.

Maybe I should update this example.

> So what will happen if you use (I did not test):
> 
> `vng echo --ssh-client test`

No, you cannot set these envs var in different places (even with `vng echo --ssh-client -- test`)

> Update: Ideally, we can get rid of `_format_cmd` or at least it should make use of `shlex.split(...)` instead of `cmd.split()`.

I didn't know about `shlex.split()`, good idea. If I'm not mistaken, `_format_cmd()` should be safe now, but better be safe than sorry.

mhartmay commented on virtme_ng/run.py:
```diff
@@ -874,23 +870,20 @@ def make(self, args):
+            make_command += [f"HOSTCC={args.compiler}", f"CC={args.compiler}"]
         if args.skip_modules:
-            make_command += f" {target}"
+            make_command.append(target)
```
Small nit and not related to your PR, but why is the order as it is? Normally, I would expect, first the target and then the makefile variables.

matttbe replied:
Note sure, but I think it doesn't matter with the Makefile, you can run `make FOO=1 target -j48 BAR=2` I think

mhartmay replied:
It's easier to read IMO :) I would expect the arguments to be sorted a little bit at least... But yes, both should work.

matttbe replied:
Sure, I can change that.

mhartmay commented on README.md:
````diff
@@ -222,9 +222,15 @@ Examples
 
  - Build the kernel using a separate build directory, and run it, in verbose:
 ```
-   $ mkdir -p .virtme/build
````
Is the directory creation no longer necessary?

matttbe replied:
good point, I forgot to mention that in the commit message. I think after a recent change on the `virtme-configkernel` side, it is no longer needed to create this dir. I will update the commit message

mhartmay commented on virtme_ng/run.py:
```diff
@@ -733,33 +731,31 @@ def checkout(self, args):
         # Propagate additional Makefile variables
-        for var in args.envs:
-            cmd += f" {var} "
+        if args.envs is not None:
```
Is None ever possible?

The original code would also fail in case of `args.envs == None`.
```
for var in None:
```
raises an exception:
`TypeError: 'NoneType' object is not iterable`

matttbe replied:
Arf, I changed this in my last version, and forgot to commit. Indeed, the condition is useless, `args.envs` is always set, but can be empty. I will remove this line and another one.

mhartmay commented on virtme_ng/run.py:
```diff
@@ -1186,7 +1189,7 @@ def _get_virtme_append(self, args):
         append = []
         if args.append is not None:
             for item in args.append:
-                split_items = item.split()
+                split_items = shlex.split(item)
                 for split_item in split_items:
                     append.append("-a " + split_item)
```
Can you please change the whole function.... it looks very suspicious to me.

```
        append = []
        if args.append is not None:
            for item in args.append:
                split_items = shlex.split(item)
                for split_item in split_items:
                   append += ["-a", split_item]
        if args.debug:
            append + = ["-a", "nokaslr"]
        self.virtme_param["append"] = shlex.join(append)
```

Still not beautiful and untested.

mhartmay replied:
And might be a good idea to use the long option for better readability (although `--kopt` reads not that good :D).

matttbe replied:
Good idea! Will do!

mhartmay commented:
LGTM, although untested.

arighi reviewed (approved):
Thanks @matttbe for working at this and sorry for the late review (I've been swamped with other proejcts).

Reviewed & tested, all looks good to me!

matttbe commented:
> Reviewed & tested, all looks good to me!

There was no hurry, thank you for the review and the testing. There are many different cases around these options, more testing is always good to have 🙂

arighi commented:
BTW @matttbe @mhartmay are you ok to cut a new release of vng? Do you have any other pending changes? Everything seems to work fine on my side with this applied, I've verified all my usual test cases. :)

mhartmay commented:
> BTW @matttbe @mhartmay are you ok to cut a new release of vng? Do you have any other pending changes? Everything seems to work fine on my side with this applied, I've verified all my usual test cases. :)

Fine with me :) 

And a quick test has shown, that

```
vng O=.virtme/build CC='ccache gcc'  --build
```
works now.

matttbe commented:
> BTW @matttbe @mhartmay are you ok to cut a new release of vng? Do you have any other pending changes? Everything seems to work fine on my side with this applied, I've verified all my usual test cases. :)

Fine by me, no other pending changes. Thank you for having checked. Good idea to have a new release now!

matttbe commented:
@arighi thank you for the new release. I noticed you reverted commit eb30683799c782d3b0362c0a66b4970bcf77f704 ("vng: exec: quote envs args"). By chance, do you have the command that was executed by the CI?

Note that as explained in the commit message of the reverted commit, it was changing the behaviour in some cases because the previous one was requiring workarounds, e.g. adding extra backslashes where it shouldn't. Maybe we can also handle these workarounds.

mhartmay commented on virtme_ng/run.py:
```diff
@@ -909,7 +903,7 @@ def _get_virtme_exec(self, args):
-            args.exec = " ".join(envs)
+            args.exec = shlex.join(envs)
         if args.exec is not None:
             self.virtme_param["exec"] = f"--script-sh {shlex.quote(args.exec)}"
```
@matttbe So with your change in https://github.com/arighi/virtme-ng/pull/288/files#diff-729892ba898a62ab180cb4ca44413dd5a538d12fb842bee5d34520e05bf042bdR908 I guess we can remove the `shlex.quote` here and hopefully this fixes the problem from @arighi .

matttbe replied:
@mhartmay thank you for looking at that.

I think we still need the `shlex.quote()` or simple/double quotes to go with `--script-sh`, no?

    --script-sh "echo bla"

Without `shlex.quote()`, we would execute:

    --script-sh echo bla

no?

My fear is that some people tried to use `\` to escape some quotes, and it no longer works. Or indeed, maybe they are using ";" to split commands.

mhartmay replied:
> @mhartmay thank you for looking at that.
> 
> I think we still need the `shlex.quote()` or simple/double quotes to go with `--script-sh`, no?
> 
> ```
> --script-sh "echo bla"
> ```
> 
> Without `shlex.quote()`, we would execute:
> 
> ```
> --script-sh echo bla
> ```
> 
> no?

If this line ` args.exec = shlex.join(envs)` is used for the assignment, then `shlex.quote` is not needed IMO. But if there is another way how `args.exec` can be set then we have to use `shlex.quote`.

So something like
```
if envs:
    args.exec = shlex.join(envs)
else:
   args.exec = shlex.quote(args.exec)
```

Note: I've not checked how args.exec can be set - so please fix the code above accordingly :)

> 
> My fear is that some people tried to use `\` to escape some quotes, and it no longer works. Or indeed, maybe they are using `;` to split commands, but then not sure how.

If they use `;` to split commands, this should never have been worked... because for this you have to set `shell=True` in `Popen` and co. See [1].

If someone needs to run multiple commands and using pipeing and so on, they have to use (untested)
```
--script-sh "sh -c  'echo first; echo second'"
```

[1] https://docs.python.org/3/library/subprocess.html#subprocess.run

matttbe replied:
> > @mhartmay thank you for looking at that.
> > I think we still need the `shlex.quote()` or simple/double quotes to go with `--script-sh`, no?
> > ```
> > --script-sh "echo bla"
> > ```
> > 
> > Without `shlex.quote()`, we would execute:
> > ```
> > --script-sh echo bla
> > ```
> > 
> > no?
> 
> If this line ` args.exec = shlex.join(envs)` is used for the assignment, then `shlex.quote` is not needed IMO. But if there is another way how `args.exec` can be set then we have to use `shlex.quote`.
> 
> So something like
> 
> ```
> if envs:
>     args.exec = shlex.join(envs)
> else:
>    args.exec = shlex.quote(args.exec)
> ```
> 
> Note: I've not checked how args.exec can be set - so please fix the code above accordingly :)

`args.envs` is an array containing all non consumed parameters, while `args.exec` is a string: e.g.

```
$ vng -r -- echo blah
args.envs == ['echo', 'blah']
$ vng -r --exec "echo blah"
args.exec == echo blah
```

If `args.envs` is set, we convert it as a string, and replace `args.exec`:

```
$ vng -r -- echo blah
args.envs == ['echo', 'blah']
args.exec == echo blah
```

We then need to pass this string as argument to `--script-sh`, and this value needs to be quoted because of the possible spaces in `args.exec`. So we need `shlex.quote()` or simple/double quotes around `args.exec`, no?

So one of these lines, but I guess we prefer the first one:

```
self.virtme_param["exec"] = f"--script-sh {shlex.quote(args.exec)}"
self.virtme_param["exec"] = f"--script-sh '{args.exec}'"
```

> > My fear is that some people tried to use `\` to escape some quotes, and it no longer works. Or indeed, maybe they are using `;` to split commands, but then not sure how.
> 
> If they use `;` to split commands, this should never have been worked... because for this you have to set `shell=True` in `Popen` and co. See [1].

Good point!

> If someone needs to run multiple commands and using pipeing and so on, they have to use (untested)
> 
> ```
> --script-sh "sh -c  'echo first; echo second'"
> ```

Indeed. It might be good to know what command stopped working then :)

mhartmay replied:
> note that if before (or now, after the revert) we were passing:

You mean version 1.35?

> 
> ```
> $ vng -r -- sh -c 'echo first; echo second'
> ```
> 
> it would do:
> 
> ```
> $ vng -r -- sh -c 'echo first; echo second'
> args.envs == ['sh', '-c', 'echo first; echo second']
> args.exec == sh -c echo first; echo second
> --script-sh 'sh -c echo first; echo second'
> ```
> 
> Instead of:
> 
> ```
> $ vng -r -- sh -c 'echo first; echo second'
> args.envs == ['sh', '-c', 'echo first; echo second']
> args.exec == sh -c 'echo first; echo second'
> --script-sh 'sh -c '"'"'echo first; echo second'"'"''
> ```
> 
> Which fixes stuff.

What? :) I'm a little confused...

> I guess when people wanted to use this command, they added extra quotes:
> 
> ```
> $ vng -r -- sh -c '"echo first; echo second"'
> ```

That looks really confusing to me.

> 
> Which would be a good workaround with `" ".join()`:
> 
> ```
> print(f"--script-sh {shlex.quote(" ".join(['sh', '-c', '"echo first; echo second"']))}")
> --script-sh 'sh -c "echo first; echo second"'
> ```
> 
> But not with `shlex.join()`:
> 
> ```
> print(f"--script-sh {shlex.quote(shlex.join(['sh', '-c', '"echo first; echo second"']))}")
> --script-sh 'sh -c '"'"'"echo first; echo second"'"'"''
> ```
> 
> But `sh -c '"echo first; echo second"'` is not a correct command. So as mentioned in the [commit](https://github.com/arighi/virtme-ng/commit/eb30683799c782d3b0362c0a66b4970bcf77f704) message, there is a behaviour change, but for a good reason. No?

Some quick tests:

```
$ git checkout v1.34
./vng -r -- sh -c 'echo first; uname -r; echo separator; hostname | grep --only-matching me'
first
6.8.0-58-generic
separator
me
$ git checkout v1.35
./vng -r -- sh -c 'echo first; uname -r; echo separator; hostname'

6.8.0-58-generic
separator
me
```
So this is broken in v1.35...

```
$ git checkout v1.34
./vng -r -- 'echo first; uname -r; echo separator; hostname | grep --only-matching me'
first
6.8.0-58-generic
separator
me
```
...but not in v1.34.

```
$ git checkout v1.34
./vng -r -- 'echo first; uname -r; echo separator; hostname | grep --only-matching me'
/run/tmp/.virtme-script: line 1: echo first; uname -r; echo separator; hostname | grep --only-matching me: command not found
```
So this was broken in v1.34...

And something that was and is currently broken....
```
$ ./vng -r --exec 'echo hello' -- uname -r
6.8.0-58-generic
```

So we should either disallow the combination of `--exec` and positional arguments or fix it. I tend to disallow it.

Next interesting scenario:
```
./vng -r --exec 'echo hello' --exec 'echo bli'
bli
````
But I guess, that's as expected.

mhartmay commented on virtme_ng/run.py:
```diff
@@ -1113,6 +1107,14 @@ def _get_virtme_remote_cmd(self, args):
+            or args.console is not None
+        ):
+            self.virtme_param["remote_cmd"] = (
+                f"--remote-cmd {shlex.quote(shlex.join(args.envs))}"
```
A quick test has shown:

```
$ python3
>>> from shlex import join
>>> print(join(['echo', '-n', 'Multiple words', ';', 'echo', 'bla']))`
echo -n 'Multiple words' ';' echo bla
$ echo -n 'Multiple words' ';' echo bla
Multiple words ; echo bla% 
```
So IMO it seems like there is no need (and we must not) use `shlex.quote` on the result of `shlex.join(...)`.

arighi commented:
> @arighi thank you for the new release. I noticed you reverted commit [eb30683](https://github.com/arighi/virtme-ng/commit/eb30683799c782d3b0362c0a66b4970bcf77f704) ("vng: exec: quote envs args"). By chance, do you have the command that was executed by the CI?
> 
> Note that as explained in the commit message of the reverted commit, it was changing the behaviour in some cases because the previous one was requiring workarounds, e.g. adding extra backslashes where it shouldn't. Maybe we can also handle these workarounds.

Hi @matttbe , sorry for reverting without giving too much details, the new release completely broke the sched_ext CI and I had to come up with a quick fix...

The command that broke is this one: 
```
 vng -v -- 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
```

I think the problem is with the backticks (which is a terrible syntax and we should use the one supported by stress-ng, passing 0, I'll fix this separately in the sched_ext CI), but in any case we shouldn't break a command that was previously working, so ideally we should re-apply your fix and support also this syntax somehow. I'll do some tests later today.

mhartmay commented:
@matttbe @arighi 
Can you please test https://github.com/mhartmay/virtme-ng/tree/quote-fixes

arighi commented:
> @matttbe @arighi Can you please test https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> 
> I need to check some other places, but looks better to me :)

Nope, this one doesn't work:
```
$ vng -vr -- echo a b
usage: virtme-run [-h] [--installed-kernel [VERSION] | --kimg [KIMG] | --kdir KDIR] --mods {none,use,auto} [-a KOPT] [--root ROOT] [--rw] [--gdb]
                  [--graphics [BINARY]] [--verbose] [--net [NET]] [--net-mac-address NET_MAC_ADDRESS] [--balloon] [--sound] [--vmcoreinfo]
                  [--snaps] [--disk NAME=PATH] [--blk-disk NAME=PATH] [--memory MEMORY] [--numa NUMA] [--numa-distance NUMA_DISTANCE]
                  [--cpus CPUS] [--name NAME] [--user USER] [--script-sh SHELL_COMMAND] [--script-exec BINARY] [--arch ARCHITECTURE]
                  [--cross-compile CROSS_COMPILE_PREFIX] [--busybox PATH_TO_BUSYBOX] [--qemu-bin QEMU_BIN] [-q QEMU_OPT] [--qemu-opts ...]
                  [--disable-microvm] [--disable-kvm] [--force-initramfs] [--force-9p] [--dry-run] [--show-command]
                  [--save-initramfs SAVE_INITRAMFS] [--show-boot-console] [--no-virtme-ng-init] [--disable-monitor] [--pwd | --cwd CWD]
                  [--rwdir RWDIR] [--rodir RODIR] [--overlay-rwdir OVERLAY_RWDIR] [--nvgpu NVGPU] [--server [{console,ssh}]]
                  [--client [{console,ssh}]] [--port PORT] [--remote-cmd COMMAND] [--ssh-tcp]
virtme-run: error: unrecognized arguments: a b
```

matttbe commented:
> > @matttbe @arighi Can you please test https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> > I need to check some other places, but looks better to me :)
> 
> Nope, this one doesn't work:
> 
> ```
> $ vng -vr -- echo a b
> usage: virtme-run [-h] [--installed-kernel [VERSION] | --kimg [KIMG] | --kdir KDIR] --mods {none,use,auto} [-a KOPT] [--root ROOT] [--rw] [--gdb]
>                   [--graphics [BINARY]] [--verbose] [--net [NET]] [--net-mac-address NET_MAC_ADDRESS] [--balloon] [--sound] [--vmcoreinfo]
>                   [--snaps] [--disk NAME=PATH] [--blk-disk NAME=PATH] [--memory MEMORY] [--numa NUMA] [--numa-distance NUMA_DISTANCE]
>                   [--cpus CPUS] [--name NAME] [--user USER] [--script-sh SHELL_COMMAND] [--script-exec BINARY] [--arch ARCHITECTURE]
>                   [--cross-compile CROSS_COMPILE_PREFIX] [--busybox PATH_TO_BUSYBOX] [--qemu-bin QEMU_BIN] [-q QEMU_OPT] [--qemu-opts ...]
>                   [--disable-microvm] [--disable-kvm] [--force-initramfs] [--force-9p] [--dry-run] [--show-command]
>                   [--save-initramfs SAVE_INITRAMFS] [--show-boot-console] [--no-virtme-ng-init] [--disable-monitor] [--pwd | --cwd CWD]
>                   [--rwdir RWDIR] [--rodir RODIR] [--overlay-rwdir OVERLAY_RWDIR] [--nvgpu NVGPU] [--server [{console,ssh}]]
>                   [--client [{console,ssh}]] [--port PORT] [--remote-cmd COMMAND] [--ssh-tcp]
> virtme-run: error: unrecognized arguments: a b
> ```

Indeed, we still need the quotes. Here it is calling:

```
virtme-run (...) --script-sh echo a b
```

(instead of `--script-sh "echo a b"`). That's what I was trying to explain above :)

mhartmay commented:
Hmm, I've updated my branch:
```
./vng -r --exec 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
```
This works, but this does not and IMO, that's the correct behavior.

For the positional arguments syntax you would have to use:
```
./vng -r -- sh -c 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
```

matttbe commented:
> > @arighi thank you for the new release. I noticed you reverted commit [eb30683](https://github.com/arighi/virtme-ng/commit/eb30683799c782d3b0362c0a66b4970bcf77f704) ("vng: exec: quote envs args"). By chance, do you have the command that was executed by the CI?
> > Note that as explained in the commit message of the reverted commit, it was changing the behaviour in some cases because the previous one was requiring workarounds, e.g. adding extra backslashes where it shouldn't. Maybe we can also handle these workarounds.
> 
> Hi @matttbe , sorry for reverting without giving too much details, the new release completely broke the sched_ext CI and I had to come up with a quick fix...

I understand this rushed action due to stress (... `-ng` :) )

> The command that broke is this one:
> 
> ```
>  vng -v -- 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
> ```
> 
> I think the problem is with the backticks (which is a terrible syntax and we should use the one supported by stress-ng, passing 0, I'll fix this separately in the sched_ext CI), but in any case we shouldn't break a command that was previously working, so ideally we should re-apply your fix and support also this syntax somehow. I'll do some tests later today.

I guess the issue is the simple quotes, right? `shlex.join()` would keep the quotes around the whole command.

> Hm... since shlex.join() produces a quoted string, maybe we just need to strip the quote. This one seems to work for me:
> 
> ```
> diff --git a/virtme_ng/run.py b/virtme_ng/run.py
> index 8d21e05..bbd466d 100644
> --- a/virtme_ng/run.py
> +++ b/virtme_ng/run.py
> @@ -903,7 +903,7 @@ class KernelSource:
>              else:
>                  envs.append(var)
>          if envs:
> -            args.exec = shlex.join(envs)
> +            args.exec = shlex.join(envs).strip("'")
>          if args.exec is not None:
>              self.virtme_param["exec"] = f"--script-sh {shlex.quote(args.exec)}"
>          else:
> ```

I think that will break other things. What about: if `args.exec` has one item, plus starts and finishes with a simple/double quotes, then we strip them?

I understand we don't want to break stuff, but the behaviour change seems needed, and also fix stuff :)
What I mean is that it seems difficult to fix all workarounds, e.g. `vng -r -- sh -c '"echo first; echo second"'` → 2 times quotes as a workaround to keep quotes... What about communicating that the behaviour has been changed/fixed in the release?

matttbe commented:
> Hmm, I've updated my branch:

Back to 1.34 with an extra commit to prevent `--exec` + `--`?
> 
> ```
> ./vng -r --exec 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
> ```
> 
> This works, but this does not:
> 
> ```
> ./vng -r -- 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
> ```
> 
> and IMO, that's the correct behavior.

Indeed. But we can do something to prevent "regression".

> For the positional arguments syntax you would have to use:
> 
> ```
> ./vng -r -- sh -c 'stress-ng -t 31 --aggressive -M -c `nproc` -f `nproc` --affinity 1 --affinity-delay 1 --affinity-pin'
> ```

Or I guess this, no?

```
./vng -r -- stress-ng -t 31 --aggressive -M -c '`nproc`' -f '`nproc`' --affinity 1 --affinity-delay 1 --affinity-pin
```

But another workaround :)

mhartmay commented:
@matttbe @arighi 
Can you please test my latest branch
https://github.com/mhartmay/virtme-ng/tree/quote-fixes

I've added this workaround...
https://github.com/mhartmay/virtme-ng/commit/87173e664f5520fd74ad2ace0f314ce68e71fe2d

So no, hopefully everything works :D

matttbe commented:
This should be enough, no?

```
diff --git a/virtme_ng/run.py b/virtme_ng/run.py
index 46e3511..8550b1e 100644
--- a/virtme_ng/run.py
+++ b/virtme_ng/run.py
@@ -909,7 +909,10 @@ class KernelSource:
             else:
                 envs.append(var)
         if envs:
-            args.exec = " ".join(envs)
+            if len(envs) == 1:
+                args.exec = envs[0]
+            else:
+                args.exec = shlex.join(envs)
         if args.exec is not None:
             self.virtme_param["exec"] = f"--script-sh {shlex.quote(args.exec)}"
         else:
```

matttbe commented:
> @matttbe @arighi Can you please test my latest branch https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> 
> I've added this workaround... [mhartmay@87173e6](https://github.com/mhartmay/virtme-ng/commit/87173e664f5520fd74ad2ace0f314ce68e71fe2d)
> 
> So now, hopefully everything works :D

hehe, same as mine. Indeed. That seems OK.

When I posted my diff, I was thinking about doing that in one line:

    args.exec = envs[0] if len(envs) == 1 else shlex.join(envs)

matttbe commented:
> > > @matttbe @arighi Can you please test my latest branch https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> > > I've added this workaround... [mhartmay@87173e6](https://github.com/mhartmay/virtme-ng/commit/87173e664f5520fd74ad2ace0f314ce68e71fe2d)
> > > So now, hopefully everything works :D
> > 
> > 
> > Can you add this (or something similar)?
> > ```
> > diff --git a/virtme_ng/run.py b/virtme_ng/run.py
> > index 6a77467..0e142fb 100644
> > --- a/virtme_ng/run.py
> > +++ b/virtme_ng/run.py
> > @@ -910,6 +910,8 @@ class KernelSource:
> >                  args.exec = envs[0]
> >              else:
> >                  args.exec = shlex.join(envs)
> > +            if not args.exec:
> > +                args.exec = None
> > ```
> 
> I would rather set the default of --exec to `None`, but yep.
> 
> [...snip]

it is set to `None` by default, no? But `args.exec` could be a list with an empty string (`['']`) I think, no?

matttbe commented:
> > > > > @matttbe @arighi Can you please test my latest branch https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> > > > > I've added this workaround... [mhartmay@87173e6](https://github.com/mhartmay/virtme-ng/commit/87173e664f5520fd74ad2ace0f314ce68e71fe2d)
> > > > > So now, hopefully everything works :D
> > > > 
> > > > 
> > > > Can you add this (or something similar)?
> > > > ```
> > > > diff --git a/virtme_ng/run.py b/virtme_ng/run.py
> > > > index 6a77467..0e142fb 100644
> > > > --- a/virtme_ng/run.py
> > > > +++ b/virtme_ng/run.py
> > > > @@ -910,6 +910,8 @@ class KernelSource:
> > > >                  args.exec = envs[0]
> > > >              else:
> > > >                  args.exec = shlex.join(envs)
> > > > +            if not args.exec:
> > > > +                args.exec = None
> > > > ```
> > > 
> > > 
> > > I would rather set the default of --exec to `None`, but yep.
> > > [...snip]
> > 
> > 
> > it is set to `None` by default, no? But `args.exec` could be a list with an empty string (`['']`) I think, no?
> 
> Yep, you're right with `None`, and no `--exec` is not a list, see
> 
> ```
> ./vng -r --exec 'echo hello' --exec 'echo bli'
> bli
> ```

sorry, I wanted to mention `envs`, not `exec`. The names are too similar :)

mhartmay commented:
> > > > > > @matttbe @arighi Can you please test my latest branch https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> > > > > > I've added this workaround... [mhartmay@87173e6](https://github.com/mhartmay/virtme-ng/commit/87173e664f5520fd74ad2ace0f314ce68e71fe2d)
> > > > > > So now, hopefully everything works :D
> > > > > 
> > > > > 
> > > > > Can you add this (or something similar)?
> > > > > ```
> > > > > diff --git a/virtme_ng/run.py b/virtme_ng/run.py
> > > > > index 6a77467..0e142fb 100644
> > > > > --- a/virtme_ng/run.py
> > > > > +++ b/virtme_ng/run.py
> > > > > @@ -910,6 +910,8 @@ class KernelSource:
> > > > >                  args.exec = envs[0]
> > > > >              else:
> > > > >                  args.exec = shlex.join(envs)
> > > > > +            if not args.exec:
> > > > > +                args.exec = None
> > > > > ```
> > > > 
> > > > 
> > > > I would rather set the default of --exec to `None`, but yep.
> > > > [...snip]
> > > 
> > > 
> > > it is set to `None` by default, no? But `args.exec` could be a list with an empty string (`['']`) I think, no?
> > 
> > 
> > Yep, you're right with `None`, and no `--exec` is not a list, see
> > ```
> > ./vng -r --exec 'echo hello' --exec 'echo bli'
> > bli
> > ```
> 
> sorry, I wanted to mention `envs`, not `exec`. The names are too similar :)

You mean this:
```
$./vng -r -- ''
qemu-system-s390x: -chardev file,id=console,path=/dev/null: Duplicate ID 'console' for chardev
```?

matttbe commented:
> > > > > > > @matttbe @arighi Can you please test my latest branch https://github.com/mhartmay/virtme-ng/tree/quote-fixes
> > > > > > > I've added this workaround... [mhartmay@87173e6](https://github.com/mhartmay/virtme-ng/commit/87173e664f5520fd74ad2ace0f314ce68e71fe2d)
> > > > > > > So now, hopefully everything works :D
> > > > > > 
> > > > > > 
> > > > > > Can you add this (or something similar)?
> > > > > > ```
> > > > > > diff --git a/virtme_ng/run.py b/virtme_ng/run.py
> > > > > > index 6a77467..0e142fb 100644
> > > > > > --- a/virtme_ng/run.py
> > > > > > +++ b/virtme_ng/run.py
> > > > > > @@ -910,6 +910,8 @@ class KernelSource:
> > > > > >                  args.exec = envs[0]
> > > > > >              else:
> > > > > >                  args.exec = shlex.join(envs)
> > > > > > +            if not args.exec:
> > > > > > +                args.exec = None
> > > > > > ```
> > > > > 
> > > > > 
> > > > > I would rather set the default of --exec to `None`, but yep.
> > > > > [...snip]
> > > > 
> > > > 
> > > > it is set to `None` by default, no? But `args.exec` could be a list with an empty string (`['']`) I think, no?
> > > 
> > > 
> > > Yep, you're right with `None`, and no `--exec` is not a list, see
> > > ```
> > > ./vng -r --exec 'echo hello' --exec 'echo bli'
> > > bli
> > > ```
> > 
> > 
> > sorry, I wanted to mention `envs`, not `exec`. The names are too similar :)
> 
> You mean this:
> 
> ```
> $./vng -r -- ''
> qemu-system-s390x: -chardev file,id=console,path=/dev/null: Duplicate ID 'console' for chardev
> ```
> 
> ?

Yes, I guess that's what @arighi was showing in https://github.com/arighi/virtme-ng/pull/288#issuecomment-2851005682

The fix could be simply this:

```
diff --git a/virtme_ng/run.py b/virtme_ng/run.py
index 46e3511..c079e89 100644
--- a/virtme_ng/run.py
+++ b/virtme_ng/run.py
@@ -910,8 +910,8 @@ class KernelSource:
                 envs.append(var)
         if envs:
             args.exec = " ".join(envs)
-        if args.exec is not None:
+        if args.exec:
             self.virtme_param["exec"] = f"--script-sh {shlex.quote(args.exec)}"
         else:
             self.virtme_param["exec"] = ""
```

mhartmay commented:
@matttbe @arighi Thanks for the feedback, I've created a PR https://github.com/arighi/virtme-ng/pull/291
